### PR TITLE
working on arg mode

### PR DESCRIPTION
**Pull request by kurtbrose**

### CODE REVIEW

**Comment by kurtbrose** [comment]
I think this is clearly on the right track -- `Call()` was already basically doing arg-mode, but doing it with a bunch of manual tap-dancing

image

**Comment by mahmoud** [comment]
Nice, yeah, there are a variety of precedents set in the existing specs. Invoke doesn't do that, though, does it? We ended up going with a more explicit approach because of the discomfort with Call's handling of this?

And for anyone following along (as well as for my own recollection), here's the bright line:

> Should "subspec"-style arguments to specifier types respect the mode of the surrounding spec? Or should they be consistent across all specs and contexts?

This probably doesn't need to be democratized, and can be determined by comparing before-and-after with some existing cases, hence this PR.

**Comment by kurtbrose** [comment]
To unpack it a bit more -- _in general_ subspecs should clearly follow Mode; the proposal here isn't to change that.

However, currently there are three kinds of values that give you different behavior:

* T, S, etc
* other specs
* non-spec python values

Depending on what you are doing "clearly" one or the other is the correct behavior.  There's a lot of work trying to thread that needle:

* special cases for `T` in `t_eval`, `assign`, etc
* `Literal()` -> `Val()`
* `Spec()` spec
* `FILL` mode
* `SKIP` and `STOP` as literals

So, this is the latest in what has been a continuous attempt to sand down a rough edge.

What is the scope of applicability?  These are some guidelines:

* specs and literal vals would both be useful
* there is special casing for `T`
* `SKIP` and `STOP` are used

These three criteria often overlap -- as was the case with `Call()`.  

As a simple motivating example:

```python
S(a=T)  # stores current Target in S.a
S(a=(T, 1))  # stores literal tuple (T, 1) in S.a
S(a=1)  # error: 1 is not a valid spec
````

with arg-mode

```python
S(T)  # stores Target in S.a
S((T, 1)) # stores (Target, 1) in S.a
S(1)  # stores 1 in S.a
```

**Comment by kurtbrose** [comment]
What are the downsides?

* forces mode understanding earlier -- I'm hopeful the examples will be so intuitive the most users won't really have to think about it, but this does mean that e.g. tuple, dicts mean different things in different parts of a spec even without explicit mode usage
* the worst case above is when using a spec that itself has subspecs, e.g. `T[Coalesce('a', default='b')]` -- since `Coalesce` would be called in `arg mode`, `'a'` will evaluate to the literal `'a'` instead of doing attribute access; however, currently that will do `Target[Coalesce(...)]`, just taking the whole Spec as a literal and trying to fetch it as a key out of the target, which I think is clearly worse
* now for every subspec we need to decide if they are an "arg-mode" type subspec or a regular subspec -- but I think in practice going through the code base it will be pretty clear


The goal is making fewer special-case cutouts for `T` by having this new mode, and ending up with an overall simpler, more consistent, more comprehensible, more powerful system.

There's also some interesting interactions with `M`.  We pulled `M(subspec)` and restricted it to only `T`.  So, this is another special case of `T`.  I don't think it is a good use case for arg-mode, because `M()` checks for truthiness -- which for all the data structures handled by arg-more will mean "is it empty", which will either be always true or always false.

**Comment by kurtbrose** [comment]
some more funny cases of arg-mode vs current behavior

```python
S(a=[])  # in arg-mode, stores a new empty list each time; currently, errors
S(a=Val([]))  # stores the same empty list every time -- unlike arg-mode has same problem as python defaults
S(a=[T])  # in arg mode, stores [Target]; in Auto mode, if Target is iterable, stores list(Target)
S(a=list)  # in arg mode, stores `list` type; in Auto, list(Target)
```

**Comment by kurtbrose** [comment]
So, tentatively I think I've found the places that can benefit:

* S()/Let() assignment (side note -- are we deprecating Let?)
* default= on Coalesce, Check, Match, Optional, Switch, And/Or
* `T[]`, `T()`, `T.`

On balance, I think the changes to the API make things simpler.

The biggest potential source of confusion is str-as-constant vs str-as-spec in the case of `S()` -- since this feature has only been out for ~ 6 weeks, I think maybe we can get out ahead of it and change the behavior without ruffling too many feathers.

**Comment by kurtbrose** [comment]
definitely the changes that were required to unit tests are interesting

**Comment by kurtbrose** [comment]
oops I missed `Assign()` -- the val argument there would probably benefit from arg-mode

THAT API has been around for a long time, so changing the behavior of `Assign(Spec(...))` is probably a bad idea; will need to think about how to allow `arg_val(Spec())` to NOT change the Mode for child specs; `Group` has a similar challenge where ideally the Mode would project down through constants, but when hitting a Spec class would revert to the outer enclosing Mode; I didn't pursue it further with `Group` b/c it seems to be a rare use case -- with it popping up where maybe I can think of a generic solution

however if you look here:

```python
        if type(self.val) is Spec:
            val = scope[glom](target, self.val, scope)
        else:
            val = self.val
```

**Comment by kurtbrose** [comment]
auto mode dict keys could also benefit (another place of T special casing)

**Comment by kurtbrose** [comment]
found kind of an interesting corner case with trace rendering; locally all tests are green now

I think it is ready to go

**Comment by kurtbrose** [comment]
maybe I should add stuff to the docs though?

**Review by mahmoud — APPROVED** [review]
I think this sort of consistency is good. We should draft some docs and release note verbiage maybe. Not sure if that goes in the PR, but seems like a good enough place.

File: glom/test/test_mutation.py [hunk 1]
```diff
@@ -38,6 +38,8 @@ def r():
 def test_assign_spec_val():
     output = glom({'b': 'c'}, Assign('a', Spec('b')))
     assert output['a'] == output['b'] == 'c'
+    output = glom({'b': 'c'}, Assign('a', Or('d', 'b')))
```
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/test/test_mutation.py:41]: that's a good point; I was thinking about how to handle this and went through some ideas:

* capture constants into an {id(val): val} dict, to avoid re-walking the same data
* however "constant" to the glom eval means no spec, which isn't truly constant, so maybe copy.deepcopy
* for both this, and finding self-reference, a "compilation" step would be helpful; since this is a min-mode, we don't have the problem of passing through intermediate specs, so an explicit `Arg()` (used in all the other specs) could do constant folding as well as self-reference detection

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/test/test_mutation.py:41]: So, the job of CompileArg would be:
* replace python-constant trees (tuples, ints, floats, strings) with Val()
* replace constant trees with a spec Copy() or similar that does copy.deepcopy
* for others, apply the existing arg mode

```python
compile_arg({
   'constant': (1, 2, 3),
   'no-specs': [1, 2, 3],
   'arg-valuate': [T, T+1, T+2]
})
# this should yield
Arg({
   'constant': Val((1, 2, 3)),
   'no-specs': Copy([1, 2, 3]),
   'arg-valuate': [T, T+1, T+2],
})
```

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/test/test_mutation.py:41]: as long as self-references are contained to Val or Copy, no problem!

if they are in the path of a sub-spec, we can blow up in the spec constructor; since the point of ArgMode is to provide a mechanism of evaluating specs at all, this seems like a pure win

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/test/test_mutation.py:51]: okay, so my earlier idea opens a can of worms (see: https://github.com/mahmoud/glom/commit/e24cd300090b85fcf47bb8e97a4979f1bc9dd02a#diff-2c22213fb5b5d623d34e35549e442dc6f0221aba97b0ba261705282131006489R2565-R2636)

I think it would work, but its very complicated

I found a more straightforward way to make arg-mode recursion safe

(we can't make all specs in general recursion safe, because we don't know what the intention of user specs may be on repetition -- e.g. file.readline() or something shouldn't be memoized.  However, for arg-mode we know it can only ever recurse over itself, which means dict+list self recursion which we CAN guarantee aren't doing anything tricky like reading lines out of a file)

File: glom/core.py [hunk 2]
```diff
@@ -187,11 +190,17 @@ def __str__(self):
...
 
 
-def _unpack_stack(scope):
+def _unpack_stack(scope, only_errors=True):
     """
     convert scope to [[scope, spec, target, error, [children]]]
 
     this is a convenience method for printing stacks
+
+    only_errors=True means ignore branches which may still be hanging around]
```
Comment (mahmoud) [inline glom/core.py:199]:
```suggestion
    only_errors=True means ignore branches which may still be hanging around
```

File: glom/core.py [hunk 3]
```diff
@@ -215,6 +224,11 @@ def _unpack_stack(scope):
         cur, nxt = stack[i], stack[i + 1]
         if cur[3] == nxt[3]:
             cur[3] = None
+    if only_errors:  # trim the stack to the last error
+        # leave at least 1 to not break formatting func below
+        # TODO: make format_target_spec_trace() tolerate an "empty" stack cleanly
```
Comment (mahmoud) [inline glom/core.py:229]: when does an empty stack happen, and what's the behavior here?

File: glom/core.py [hunk 4]
```diff
@@ -2267,7 +2270,7 @@ def glom(target, spec, **kwargs):
         except skip_exc:
             if default is _MISSING:
                 raise
-            ret = default
+            ret = default  # should this also be arg_val'd?
```
Comment (mahmoud) [inline glom/core.py:2273]: if we evaluate early, how bad would it be for perf? and if we evaluate it late (i.e., at this line), and it fails, that seems like it would be bad for reliability?

File: glom/core.py [hunk 5]
```diff
@@ -112,6 +113,8 @@
 
 MODE =  make_sentinel('MODE')
 
+MIN_MODE =  make_sentinel('MIN_MODE')
```
Comment (mahmoud) [inline glom/core.py:116]: Do we describe minor modes anywhere? (comments/docs)

File: glom/core.py [hunk 6]
```diff
@@ -2531,3 +2537,38 @@ def FILL(target, spec, scope):
...
+        result = spec
+        if type(spec) in (list, dict):  # can contain themselves
+            if id(spec) in self.cache:
+                return self.cache[id(spec)]
+            result = self.cache[id(spec)] = type(spec)()
+            if type(spec) is dict:
+                result.update({recur(key): recur(val) for key, val in spec.items()})
+            else:
+                result.extend([recur(val) for val in spec])
+        if type(spec) in (tuple, set, frozenset):  # cannot contain themselves
```
Comment (mahmoud) [inline glom/core.py:2560]: Just a thought, but how do we ensure that arg mode stays supporting the same set of containers as Fill? (dict, list, tuple, set, frozenset) Also, do we expect that will grow? Is it worth documenting them separately if they might diverge in future?

File: glom/test/test_error.py [hunk 7]
```diff
@@ -60,20 +60,6 @@ def test_line_trace():
     assert fmtd_stack == '/tuple!list/list<0>/dict!int/S'
 
 
-def test_short_trace():
```
Comment (mahmoud) [inline glom/test/test_error.py:63]: curious why this one got removed.

**Review by mahmoud — COMMENTED** [review]
Just a few initial thoughts/Qs, still reviewing

File: glom/core.py [hunk 8]
```diff
@@ -2531,3 +2537,38 @@ def FILL(target, spec, scope):
...
+            result = type(spec)([recur(val) for val in spec])
+        return result
+
+
+def arg_val(target, arg, scope):
```
Comment (mahmoud) [inline glom/core.py:2565]: should `arg_val` be more documented and public? Just thinking that if others want to develop specs, then they'll conventionally want to use this for arguments?

File: glom/core.py (hunk 3 above)
Comment (kurtbrose) [reply glom/core.py:229]: the stack formatting is a bit of an ad-hoc mess right now; I think this may have been something unrelated that I just ran across

File: glom/core.py (hunk 5 above)
Comment (kurtbrose) [reply glom/core.py:116]: that's a pretty big oversight, good point

I think there may be some docstrings attached to ARG / arg_val

File: glom/core.py (hunk 6 above)
Comment (kurtbrose) [reply glom/core.py:2560]: my initial thought was "things that python has syntax support for" -- but then I went and included frozenset

Comment (kurtbrose) [reply glom/core.py:2560]: I don't expect this to grow -- its easy enough to embed your own callables into glom

in some sense, these data structures form the "syntax" of glom (nested tuples, lists, dicts)

File: glom/core.py (hunk 8 above)
Comment (kurtbrose) [reply glom/core.py:2565]: yea, that's a good point -- maybe arg_val is the right thing to make public and throw a big comment on

as I was playing around with the compilation I thought maybe it could work kind of like the re module:

arg_val = re.match

arv_compile = re.compile

so, arg_val and arg_compile would be the public entry points

in which case, maybe we should take a minute to think if arg_val is the right name: should it be arg_eval?  arg_glom?

Comment (kurtbrose) [reply glom/core.py:2565]: sorry, to round out the re parallel:

like re, you can do arg_val(arg_compile(arg)) or arg_val(arg), and they'll both give the same result

like re, its up to the owning spec to figure out where it wants to store the result of arg_compile

File: glom/core.py (hunk 4 above)
Comment (kurtbrose) [reply glom/core.py:2273]: yea, I'm inclined now to back off a little bit and leave things lay

Comment (kurtbrose) [reply glom/core.py:2273]: it was just a thought experiment -- if arg_val makes sense for all these internal spec defaults, what about the top level default?

I think you're right about reliability though, and also I think the entry point is a special case -- the entire glom "program" is meant to be contained in the spec, and I think that's an important top-level invariant

File: glom/test/test_error.py (hunk 7 above)
Comment (kurtbrose) [reply glom/test/test_error.py:63]: I think only_errors got added to _unpack_stack() trim "extra" branches, but in the process it kind of broke this test case -- it relies on extracting the scope from inside an executing glom that then completes and formatting it

this isn't really intended to be possible

also, this was giving low value vs the ones that actually use a raised exception
